File: BucketSort/BucketSort.py
import random

def randomnumber(a):
    for i in range(len(a)):
     a[i]=random.randint(1,100)
def printarray(a):
   for i in range(len(a)):
     print(a[i], end=" ")

#Algoritmo de ordenamiento Bucket Sort
def insertion_sort(bucket):
    for j in range(1, len(bucket)):
        val = bucket[j]
        k = j - 1
        while k >= 0 and bucket[k] > val:
            bucket[k + 1] = bucket[k]
            k -= 1
        bucket[k + 1] = val  

def bucket_sort(inputArr):
    s = len(inputArr)
    bucketArr = [[] for _ in range(s)]

    max_val = max(inputArr)

    # Distribuir los elementos en los buckets 
    for j in inputArr:
        normalized = j / (max_val + 1)
        bi = int(s * normalized)
        bucketArr[bi].append(j)

    # Ordenar cada bucket con insertion sort
    for bucket in bucketArr:
        insertion_sort(bucket)

    # Concatenar los buckets ordenados
    idx = 0
    for bucket in bucketArr:
        for j in bucket:
            inputArr[idx] = j
            idx += 1

array=[0]*15
randomnumber(array)
print("Arreglo Original: ")
printarray(array)
bucket_sort(array)
print("\nArray Ordenado:")
printarray(array)

File: BucketSort/BucketSort.js
function random(min, max){
    return Math.floor(Math.random()*(max-min+1))+min;
}

function printarray(a){
    for(let i=0;i<a.length;i++){
    console.log(a[i])
}
}

function insertionSort(bucket) {
    for (let j = 1; j < bucket.length; j++) {
        let key = bucket[j];
        let i = j - 1;
        while (i >= 0 && bucket[i] > key) {
            bucket[i + 1] = bucket[i];
            i = i - 1;
        }
        bucket[i + 1] = key;
    }
}
function bucketSort(inputArr) {
    let s = inputArr.length;
    let bucketArr = new Array(s);
    for (let i = 0; i < s; i++) {
        bucketArr[i] = [];
    }
    let maxVal = Math.max(...inputArr);
    // Distribuir los elementos en los buckets
    for (let j = 0; j < inputArr.length; j++) {
        let normalized = inputArr[j] / (maxVal + 1);
        let bi = Math.floor(s * normalized);
        bucketArr[bi].push(inputArr[j]);
    }
    // Ordenar cada bucket con insertion sort
    for (let i = 0; i < bucketArr.length; i++) {
        insertionSort(bucketArr[i]);
    }
    // Concatenar los buckets ordenados en el array de entrada
    let idx = 0;
    for (let i = 0; i < bucketArr.length; i++) {
        for (let j = 0; j < bucketArr[i].length; j++) {
            inputArr[idx] = bucketArr[i][j];
            idx++;
        }
    }
}

let arr=new Array(15)

for(let i=0;i<arr.length;i++){
    arr[i]=random(1,100);
}
console.log("\nArreglo Original: ");

printarray(arr)
bucketSort(arr)
console.log("\nArray Ordenado: ")
printarray(arr)
